Reduce duplication

ericpoe opened the pull request:
This is a code-cleanup PR.

This PR reduces a bunch of duplication in the Functional classes.

HStringContains and HArrayMap have been cleaned up so that the intention is easier to discern.

ericpoe committed:
Simplify HArrayMap & HStringMap ->map

Compare against interface rather than class

ericpoe committed:
Initial cleanup of ->map.

Ensures an error gets thrown if a non-array is passed in the variadic.

Both HArrayMap->map() and HStringMap->map() are exactly the same. I should
be able to minimize this duplication

ericpoe committed:
Condense two separate map classes into one

ericpoe committed:
Correct spacing

ericpoe committed:
Return from exit condition sooner

Since an array doesn't need to be converted to itself, there's
no reason to go any further in this method.

ericpoe committed:
Cleanup Exception text

ericpoe committed:
Rename Container files to the shortened form

ericpoe committed:
Cleanup HStringReduce so that it looks like HArrayReduce

This opens the door for combining both classes into one.

ericpoe committed:
Combine HArrayReduce and HStringReduce into one class

ericpoe committed:
Combine two -FilterWithDefaults classes into one

ericpoe committed:
Reduce duplication by combining two FilterWithValue classes into one

ericpoe committed:
Reduce duplication by combining two FilterWithKey classes

ericpoe committed:
Reduce duplication by combining two FilterWithValueAndKey classes

ericpoe committed:
Reduce duplication by combining two Filter classes into one

ericpoe committed:
Fix spacing

ericpoe committed:
Add helper docblock

ericpoe committed:
Fix spacing

ericpoe commented on src/Container/HStringContains.php:
```diff
@@ -0,0 +1,48 @@
+<?php
```
I don't know why git thinks this is created here and deleted later when this was just a simple rename.

There are also a few changes in the `contains` method that simplifies its behavior.

johnkary commented on src/Container/HStringContains.php:
```diff
@@ -0,0 +1,48 @@
+        if (is_scalar($value)) {
+            $this->value = (string) $value;
+        } else {
+            throw new \InvalidArgumentException(sprintf("%s is neither a scalar value nor an HString", Helper::getType($value)));
```
[NB] This error message says _what_ is wrong but doesn't describe _why_ it's wrong. What's wrong with passing in a value that's not a scalar or HString? What can't be done that was intended to be done when the developer called this method?

ericpoe replied:
It's a good thought. I'll think about an appropriate exception text.

johnkary commented on src/Container/HStringContains.php:
```diff
@@ -0,0 +1,48 @@
+    {
+        $pos = strpos($this->str, $this->value);
+
+        return (false !== $pos) ?: false;
```
[NB] I find the ternary operator use here makes this confusing. It's easy to reason about but I have to think a minute to be sure I've got it straight.

The true-condition block doesn't explicitly list a value `true`, yet has `false` in the false-condition block. The ternary isn't needed in this case.

That slight quick is compounded by the double-negative `false !==` part of the comparison. `false` ends up appearing twice, and the not-equals can trip me up. So here's my thought process when looking at this function...

"So when false is not equal to $pos... What's the $pos again?
Oh yeah the numeric position if the string is found... unless it's not found, then it's `false`, right?
OK then which part of the ternary operator are we returning...? The missing true-condition? Or wait, the false-condition? And the false-condition is `false` but wait what's the true-condition? A number from strpos() or false?
Or?
😣 
Sigh, I suck at programming."

I'd like using an instance variable to make the return statement easier to read if we were doing more complex true/false work but in the ternary operator, but in this case I think all these issues are solved by simply returning the result of comparing the strpos return value:

```
private function containsValue()
{
    return false !== strpos($this->str, $this->value);
}
```

What do you think?

ericpoe replied:
That certainly makes more sense. I was struggling to make the line more clear and ended up making it more convoluted!

johnkary commented:
All this cleanup looks great! `HStringContains` looks much more simple.

A lot of classes changed from "HaystackString*" to "HString*", also "HaystackArray*" to "HArray*". What about the remaining classes that have "Haystack" still in the title? Could these be shortened similarly?
- src/Functional/HaystackFilter.php
- src/Functional/HaystackFilterWithDefaults.php
- src/Functional/HaystackFilterWithKey.php
- src/Functional/HaystackFilterWithValue.php
- src/Functional/HaystackFilterWithValueAndKey.php
- src/Functional/HaystackMap.php
- src/Functional/HaystackReduce.php

ericpoe committed:
Simplify return statement

ericpoe committed:
Rename HaystackFilter... classes to be just Filter...

Since these classes can be used by both HArray and HString, the choice is to either keep or remove "Haystack" from the name. I'm opting for shorter classnames here. Woo!

ericpoe committed:
Allow objects with a defined `__toString` to be used as search values

ericpoe committed:
Improve error message for HString->contains()

ericpoe committed:
Clean up use statements

Minor, I know. But I like the use statements to be in alhpa order

ericpoe committed:
Remove unused USE statements

ericpoe committed:
Simplify HString::locate

ericpoe committed:
Clean up filter
